Actually use a fixed seed in `UTxOIndexBench`.

jonathanknowles opened the pull request:
## Issue Number

ADP-2975

## Summary

This PR fixes the generation of UTxO maps in `UTxOIndexBench`.

The previous method of generation assumed that it was possible to use the `QC.variant` function to generate values according to a fixed seed, but that assumption was flawed. This flaw would result in the benchmark generating different UTxO maps on every run, which could cause unwanted variance in the benchmark timings.

This PR:
- adds a new function `generateWith` to `Test.QuickCheck.Extra`.
- uses the `generateWith` function to repair the generation logic in `UTxOIndexBench`.

jonathanknowles committed:
Add function `generateWith` to `Test.QuickCheck.Extra`.

The function `generateWith` is a pure alternative to the standard
QuickCheck `generate` function, allowing the caller to specify a
particular seed and size parameter.

This can be useful for benchmarking, where it's sometimes useful to be
able to generate the exact same value over multiple runs of a benchmark.

jonathanknowles committed:
Use function `generateWith` in `UTxOIndexBench`.

This replaces the `generateWithSeed` function, which does NOT have the
advertised behaviour (unfortunately).

jonathanknowles commented on lib/primitive/bench/UTxOIndexBench.hs:
```diff
@@ -227,11 +227,6 @@ minimalTokenQuantity = TokenQuantity 1
--- | Generates values using a specific PRNG seed.
---
-generateWithSeed :: Int -> Gen a -> IO a
-generateWithSeed seed = generate . variant seed
```
This was completely wrong: `variant` does not work in the way we thought. ^^^

jonathanknowles commented on lib/primitive/bench/UTxOIndexBench.hs:
```diff
@@ -141,13 +143,11 @@ makeUTxOMapNFTs size =
+makeUTxOMapDiverse size =
+    -- We use a fixed PRNG seed and QC size parameter to maximise consistency
+    -- between benchmark runs:
+    generateWith (GenSeed 0) (GenSize 30) (genUTxOMapDiverse size)
```
`30` is the standard QC size parameter. See here:
https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/src/Test.QuickCheck.Gen.html#generate

Unisay replied:
How do you like this:
```haskell
let noRandomness = GenSeed 0
    stdQuickCheckSize = GenSize 30
in  generateWith noRandomness stdQuickCheckSize (genUTxOMapDiverse size)
```
?

jonathanknowles replied:
> How do you like this:
> ```haskell
> let noRandomness = GenSeed 0
>     stdQuickCheckSize = GenSize 30
> in  generateWith noRandomness stdQuickCheckSize (genUTxOMapDiverse size)
> ``` 
> ?

Not so sure about `generateWith noRandomness`, since the name `noRandomness` **_seems_** to imply that this particular choice of argument causes it to have no randomness. But changing the value of this argument from one value to another just changes the initial seed for the PRNG, and doesn't make it any "more random" or "less random".

Regarding the suggestion to use `stdQuickCheckSize`: I could do something to make this choice more obvious. I'll push another commit!

Unisay replied:
agree on "no randomness". My point is that we can bind names to expressions such that names explain the purpose of those magic numbers thus making them less magic.

jonathanknowles replied:
Extra commit pushed here:
8e094b3b3a1490985ab8bcb7bab3291e02a18042

Unisay commented on lib/test-utils/src/Test/QuickCheck/Extra.hs:
```diff
@@ -255,6 +264,29 @@ shrinkInterleaved (a, shrinkA) (b, shrinkB) = interleave
+--
+generateWith :: GenSeed -> GenSize -> Gen a -> a
+generateWith (GenSeed seed) (GenSize size) (MkGen runGen) =
+    runGen (mkQCGen seed) size
```
👍🏼

jonathanknowles committed:
Add `genSizeDefault`: a default value for `GenSize`.

Based on the default value of the QC size parameter (30).

jonathanknowles committed:
Rename `size` arguments to `mapSize` in `UTxOIndexBench`.

This helps to distinguish the following two notions:

- The size of a UTxO map.
- The QC size parameter.

jonathanknowles committed:
Use `genSizeDefault` in `UTxOIndexBench`.
